src/LockTable.h

```diff
@@ -0,0 +1,187 @@
+/* Copyright (c) 2014-2015 Stanford University
+ *
+ * Permission to use, copy, modify, and distribute this software for any
+ * purpose with or without fee is hereby granted, provided that the above
+ * copyright notice and this permission notice appear in all copies.
+ *
+ * THE SOFTWARE IS PROVIDED "AS IS" AND THE AUTHOR(S) DISCLAIM ALL WARRANTIES
+ * WITH REGARD TO THIS SOFTWARE INCLUDING ALL IMPLIED WARRANTIES OF
+ * MERCHANTABILITY AND FITNESS. IN NO EVENT SHALL AUTHORS BE LIABLE FOR
+ * ANY SPECIAL, DIRECT, INDIRECT, OR CONSEQUENTIAL DAMAGES OR ANY DAMAGES
+ * WHATSOEVER RESULTING FROM LOSS OF USE, DATA OR PROFITS, WHETHER IN AN
+ * ACTION OF CONTRACT, NEGLIGENCE OR OTHER TORTIOUS ACTION, ARISING OUT OF
+ * OR IN CONNECTION WITH THE USE OR PERFORMANCE OF THIS SOFTWARE.
+ */
+
+#ifndef RAMCLOUD_LOCKTABLE_H
+#define RAMCLOUD_LOCKTABLE_H
+
+#include "Common.h"
+
+#include "Atomic.h"
+#include "Fence.h"
+
+namespace RAMCloud {
+
+/**
+ * The LockTable represents a collection a locks with unique IDs.  The locks can
+ * be acquired and released like any typical lock based our their lockId. The
+ * only restriction on the 64-bit lockId is that the value 0 can not be used.
+ *
+ * This class is used, for instance, to manage per %RAMCloud object write locks
+ * that protect against concurrent write transactions and write requests on the
+ * same object.  In this case, the lockId can be the address of the object in
+ * the %RAMCloud log.
+ *
+ * This class is thread-safe.
+ *
+ * \section impl Implementation Details
+ *
+ * Locks represented in the LockTable are implemented as spin-locks.  Blocking
+ * on a lock will result in busying waiting.
+ *
+ * The LockTable is an array of #buckets.  Locks are "hashed" into buckets based
+ * on the lower n-1 bits of the lockId (n is such that 2^n is the number of
+ * buckets); the implementation assumes that lockIds are relatively uniformly
+ * distributed.
+ *
+ * Each bucket consists of a one or more chained \link CacheLine
+ * CacheLines\endlink, the first of which lives inline in the array of buckets.
+ * Each cache line consists of several LockTable Entry objects; each Entry
+ * contains the lockId of an acquired lock or 0; only acquired locks are
+ * kept in the LockTable.  In exception, the first entry in a bucket's first
```

Comment:
I'm not sure what you mean by "In exception": is this true in every bucket? If so, just change "In exception, the" => "The"